### PR TITLE
build: remove duplicate devtools sources

### PR DESCRIPTION
**Pull request by deepak1556**
##### Description of Change

Depends on https://github.com/electron/libchromiumcontent/pull/681

Directly build from chrome devtools sources for some of the classes
  * `DevtoolsContentsResizingStrategy`
  * `DevtoolsEmbedderDispatcher`
  * `DevtoolsFileSystemIndexer`
  * `SecurityStateTabHelper`

Also have created a separate target under `chromium_src` which will build chrome sources that are depended upon by Electron unconditionally, essentially chrome sources that cannot be moved behind a build flag or don't have a target in upstream for which we have source copies. The end goal is to get rid of all source copies in `chromium_src` and move them either into build time targets under `//electron/components` or build them directly from `//electron/chromium_src:chrome` with patches in libcc.

Some of the potential sources for the chrome target are (not complete) : 
  * `IconLoader`
  * `ColorChooser`
  * `ProcessSingleton`
  * `GlobalShortcutListener`

##### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

##### Release Notes

Notes: no-notes

### CODE REVIEW

File: chromium_src/BUILD.gn [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+# on unconditionally.
+source_set("chrome") {
+  visibility = [ "//electron:electron_lib" ]
+  sources = [
+    "//chrome/browser/ssl/security_state_tab_helper.cc",
+    "//chrome/browser/ssl/security_state_tab_helper.h",
+  ]
+  public_deps = [ "//skia" ]
+  deps = [ "//components/security_state/content" ]
+}
```
Comment (deepak1556) [inline chromium_src/BUILD.gn]: Will fix this newline when updating the libcc commit.

**Comment by MarshallOfSound** [comment]
> ProcessSingleton

When / if someone gets round to doing that they need to be quite careful, we've floated a few Electron specific modifications on top of the singleton logic

File: brightray/browser/inspectable_web_contents_delegate.h [hunk 2]
```diff
@@ -3,6 +3,8 @@
 
 #include <string>
 
+#include "base/files/file_path.h"
```
Comment (nornagon) [inline brightray/browser/inspectable_web_contents_delegate.h]: surprising that this wasn't needed before 😮

File: brightray/browser/inspectable_web_contents_impl.h [hunk 3]
```diff
@@ -92,7 +92,7 @@ class InspectableWebContentsImpl
   void AppendToFile(const std::string& url,
                     const std::string& content) override;
   void RequestFileSystems() override;
-  void AddFileSystem(const std::string& file_system_path) override;
+  void AddFileSystem(const std::string& type) override;
```
Comment (nornagon) [inline brightray/browser/inspectable_web_contents_impl.h]: the path is no longer needed...?

File: chromium_src/BUILD.gn [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# Use of this source code is governed by the MIT license that can be
+# found in the LICENSE file.
+
+# Builds some of the chrome sources that Electron depends
+# on unconditionally.
+source_set("chrome") {
+  visibility = [ "//electron:electron_lib" ]
+  sources = [
+    "//chrome/browser/ssl/security_state_tab_helper.cc",
+    "//chrome/browser/ssl/security_state_tab_helper.h",
```
Comment (nornagon) [inline chromium_src/BUILD.gn]: should the devtools sources from brightray also go here?

File: chromium_src/BUILD.gn [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  ]
+  public_deps = [ "//skia" ]
```
Comment (nornagon) [inline chromium_src/BUILD.gn]: why is this here?

File: brightray/browser/inspectable_web_contents_impl.h (hunk 3 above)
Comment (deepak1556) [reply brightray/browser/inspectable_web_contents_impl.h]: The path was always empty when invoked from devtools, users get to pick the folder with the file dialog the client provides and we get the path from it. The type of filesystem was added in https://chromium-review.googlesource.com/c/chromium/src/+/729250, have made the changes to reflect it. Also verified it works as expected.

File: chromium_src/BUILD.gn (hunk 4 above)
Comment (deepak1556) [reply chromium_src/BUILD.gn]: The brightray devtools sources aren't used outside of it, when we have moved the relevant `inspectable_web_contents_*` into electron, I am hoping to move the rest here. As for `SecurityStateTabHelper` its used by `atom/browser/common_web_contents_delegate` which acts as the delegate for devtools web contents, and relevant security tab info strings are obtained through it.

File: chromium_src/BUILD.gn (hunk 5 above)
Comment (deepak1556) [reply chromium_src/BUILD.gn]: Without it it errors with,

```
In file included from ../../chrome/browser/ssl/security_state_tab_helper.cc:5:
In file included from ../../chrome/browser/ssl/security_state_tab_helper.h:12:
In file included from ../../content/public/browser/web_contents_observer.h:24:
In file included from ../../third_party/skia/include/core/SkColor.h:11:
In file included from ../../third_party/skia/include/core/SkScalar.h:11:
In file included from ../../third_party/skia/include/core/../private/SkFloatingPoint.h:11:
../../third_party/skia/include/core/../private/../private/SkFloatBits.h:12:10: fatal error: 'SkTypes.h' file not found
#include "SkTypes.h"
         ^~~~~~~~~~~
1 error generated.
```

running `gn path out/Debug //electron/chromium_src:chrome //skia`, resulted in 

```
//electron/chromium_src:chrome --[private]-->
//components/security_state/content:content --[private]-->
//third_party/WebKit/public:blink_headers --[public]-->
//skia:skia
```

Identified the fix through a similar bug in upstream https://bugs.chromium.org/p/chromium/issues/detail?id=702033

File: atom/browser/common_web_contents_delegate.cc [hunk 6]
```diff
@@ -120,16 +125,18 @@ PrefService* GetPrefService(content::WebContents* web_contents) {
...
       pref_service->GetDictionary(prefs::kDevToolsFileSystemPaths);
-  std::set<std::string> result;
+  std::map<std::string, std::string> result;
   if (file_system_paths_value) {
     base::DictionaryValue::Iterator it(*file_system_paths_value);
     for (; !it.IsAtEnd(); it.Advance()) {
-      result.insert(it.key());
+      std::string type =
+          it.value().is_string() ? it.value().GetString() : std::string();
+      result[it.key()] = type;
```
Comment (ckerr) [inline atom/browser/common_web_contents_delegate.cc]: Not sure I understand the desired result here. Why do we want some entries in the map to be ""? In what conditions do we get an `it.value()` that's not a string?

File: atom/browser/common_web_contents_delegate.cc [hunk 7]
```diff
@@ -433,11 +443,12 @@ void CommonWebContentsDelegate::DevToolsIndexPath(
               file_system_path,
               base::Bind(
                   &CommonWebContentsDelegate::OnDevToolsIndexingWorkCalculated,
-                  base::Unretained(this), request_id, file_system_path),
+                  weak_factory_.GetWeakPtr(), request_id, file_system_path),
```
Comment (ckerr) [inline atom/browser/common_web_contents_delegate.cc]: :+1: reminds me of last week :smile:

File: atom/browser/common_web_contents_delegate.cc (hunk 6 above)
Comment (deepak1556) [reply atom/browser/common_web_contents_delegate.cc]: For each loaded file path devtools expects an associated file system type, when its not special an empty value must be provided.

>  In what conditions do we get an it.value() that's not a string?

`base::Value::Get*` have `CHECK`, to avoid a crash in release build due to some invalid values, the check is present as a guard. As the type value is something that is generated by the devtools rather than by a user.

File: chromium_src/BUILD.gn (hunk 5 above)
Comment (nornagon) [reply chromium_src/BUILD.gn]: hah, weird :)

Comment (nornagon) [reply chromium_src/BUILD.gn]: ... perhaps this should be `deps = [ "//content/public/browser" ]` though?

**Comment by deepak1556** [comment]
This is ready for merge, but would like to wait for https://github.com/electron/electron/pull/14564 . Will rebase on it, so that the complete test suite is run on this PR.
